src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,166 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size;
+    private Node<T> head;
+    private Node<T> tail;
+
+    public MyLinkedList() {
+        head = null;
+        tail = null;
+    }
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(value);
+        if (head == null) {
+            head = newNode;
+            tail = newNode;
+        } else {
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexForAdd(index);
+        if (index == size) {
+            add(value);
+        } else {
+            Node<T> newNode = new Node<>(value);
+            if (index == 0) {
+                newNode.next = head;
+                head.prev = newNode;
+                head = newNode;
+            } else {
+                Node<T> currentNode = getNode(index);
+                newNode.prev = currentNode.prev;
+                newNode.next = currentNode;
+                currentNode.prev.next = newNode;
+                currentNode.prev = newNode;
+            }
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = getNode(index);
+        return node.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = getNode(index);
+        T oldValue = node.value;
+        node.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
```

Comment:
Create a separate method unlink(Node node).

This logic will be used in 2 methods: when we remove from list by index and by value. So let’s create a method that will take a Node that needs to be removed and unlink it. As a result we will call it from both methods when Node is found.

For example:
```
    @Override
    public T remove(int index) {
        checkIndex(index);
        checkIndexEqualsSize(index);
        Node<T> currentNode = getCurrentNodeBy(index);
        unlinkNode(currentNode);
        return currentNode.element;
    }
```